[SwiftLexicalLookup][GSoC] New lookup method signature and `Identifier` initializer changes

MAJKFL opened the pull request:
This PR includes changes suggested by @ahoppen under #2748 that didn't make it before merging.

ahoppen commented on Sources/SwiftLexicalLookup/ScopeSyntax.swift:
```diff
@@ -74,47 +73,47 @@ extension SyntaxProtocol {
 
-  func checkName(_ name: Identifier?, refersTo introducedName: LookupName, at origin: AbsolutePosition) -> Bool {
-    introducedName.isAccessible(at: origin) && (name == nil || introducedName.refersTo(name!))
+  func checkName(_ name: Identifier?, refersTo introducedName: LookupName, at lookUpPosition: AbsolutePosition) -> Bool
+  {
```
I’d prefer to break at the paramters

```suggestion
  func checkName(
    _ name: Identifier?, 
    refersTo introducedName: LookupName, 
    at lookUpPosition: AbsolutePosition
  ) -> Bool {
```

ahoppen reviewed (commented):
Thanks for addressing my review comments. I think the following comments from https://github.com/swiftlang/swift-syntax/pull/2748 are still open.

- [ ] https://github.com/swiftlang/swift-syntax/pull/2748#discussion_r1706062399
- [ ] https://github.com/swiftlang/swift-syntax/pull/2748#discussion_r1706067597
- [ ] https://github.com/swiftlang/swift-syntax/pull/2748#discussion_r1706086462
- [ ] https://github.com/swiftlang/swift-syntax/pull/2748#discussion_r1706088559

MAJKFL commented:
1. I removed `LookupName.refersTo` and renamed `ScopeSyntax.checkName` to `ScopeSyntax.checkIdentifier` for extra clarity.
2. As far as I understand, it's not used anywhere in the compiler right now. Though there's this weird top level code behavior in playgrounds I described [here](https://github.com/swiftlang/swift-syntax/pull/2748#discussion_r1700084713). To be honest, `memberBlock` configuration wouldn't be perfect either, as it doesn't extract names from `guard`. I think everything will become clearer once we start testing against the compiler and as for now I think I'd like to leave the implementation as is so we can later iterate on it.
3. I think I might not fully understand this one. I added a test case with your example [here](https://github.com/swiftlang/swift-syntax/blob/d45771588c9fb6bd45b941df1331d150ad12f7d4/Tests/SwiftLexicalLookupTest/NameLookupTests.swift#L661-L665). As far as I understand the result [should be empty](https://github.com/swiftlang/swift-syntax/blob/d45771588c9fb6bd45b941df1331d150ad12f7d4/Tests/SwiftLexicalLookupTest/NameLookupTests.swift#L672), or am I missing something?
4. I wanted to provide extra flexibility for sequential scopes to be able to use any result kinds, but I see now it was probably superfluous. I added a new [`LookupResult.getResult`](https://github.com/swiftlang/swift-syntax/blob/2653a6e0bf7b3a577494fb1536027f36b46de9cc/Sources/SwiftLexicalLookup/LookupResult.swift#L40-L48) static method that creates a result specific for the provided scope kind.

ahoppen commented:
> 1. I removed `LookupName.refersTo` and renamed `ScopeSyntax.checkName` to `ScopeSyntax.checkIdentifier` for extra clarity.

Thanks 👍🏽 

> 2. As far as I understand, it's not used anywhere in the compiler right now. Though there's this weird top level code behavior in playgrounds I described [here](https://github.com/swiftlang/swift-syntax/pull/2748#discussion_r1700084713). To be honest, `memberBlock` configuration wouldn't be perfect either, as it doesn't extract names from `guard`. I think everything will become clearer once we start testing against the compiler and as for now I think I'd like to leave the implementation as is so we can later iterate on it.

OK, just wanted to make sure we are aware of it and don’t carry around old cruft just because nobody knows what it might be used for 👍🏽 

> 3. I think I might not fully understand this one. I added a test case with your example [here](https://github.com/swiftlang/swift-syntax/blob/d45771588c9fb6bd45b941df1331d150ad12f7d4/Tests/SwiftLexicalLookupTest/NameLookupTests.swift#L661-L665). As far as I understand the result [should be empty](https://github.com/swiftlang/swift-syntax/blob/d45771588c9fb6bd45b941df1331d150ad12f7d4/Tests/SwiftLexicalLookupTest/NameLookupTests.swift#L672), or am I missing something?

Oh sorry, I missed that test case. Looks good.

> 4. I wanted to provide extra flexibility for sequential scopes to be able to use any result kinds, but I see now it was probably superfluous. I added a new [`LookupResult.getResult`](https://github.com/swiftlang/swift-syntax/blob/2653a6e0bf7b3a577494fb1536027f36b46de9cc/Sources/SwiftLexicalLookup/LookupResult.swift#L40-L48) static method that creates a result specific for the provided scope kind.

That’s a good step. I still think that it’s worth exploring to unify the `fromFileScope` and `fromScope` cases.

ahoppen commented:
@swift-ci Please test
